pysr/julia_helpers.py

```diff
@@ -1,284 +1,34 @@
-        )
-
-    if not info.is_pycall_built():
-        raise ImportError(_import_error())
-
-    from julia.core import Julia
-
-    try:
-        Julia(**julia_kwargs)
-    except UnsupportedPythonError:
-        # Static python binary, so we turn off pre-compiled modules.
-        julia_kwargs = {**julia_kwargs, "compiled_modules": False}
-        Julia(**julia_kwargs)
-        warnings.warn(
-            "Your system's Python library is static (e.g., conda), so precompilation will be turned off. For a dynamic library, try using `pyenv` and installing with `--enable-shared`: https://github.com/pyenv/pyenv/blob/master/plugins/python-build/README.md#building-with---enable-shared."
-        )
-
-    using_compiled_modules = (not "compiled_modules" in julia_kwargs) or julia_kwargs[
-        "compiled_modules"
-    ]
-
-    from julia import Main as _Main
-
-    Main = _Main
-
-    if julia_activated_env is None:
-        julia_activated_env = processed_julia_project
 
-    if julia_initialized and julia_kwargs_at_initialization is not None:
-        # Check if the kwargs are the same as the previous initialization
-        init_set = set(julia_kwargs_at_initialization.items())
-        new_set = set(julia_kwargs.items())
-        set_diff = new_set - init_set
-        # Remove the `compiled_modules` key, since it is not a user-specified kwarg:
-        set_diff = {k: v for k, v in set_diff if k != "compiled_modules"}
-        if len(set_diff) > 0:
-            warnings.warn(
-                "Julia has already started. The new Julia options "
-                + str(set_diff)
-                + " will be ignored."
-            )
-
-    if julia_initialized and julia_activated_env != processed_julia_project:
-        Main.eval("using Pkg")
-
-        io_arg = _get_io_arg(quiet)
-        # Can't pass IO to Julia call as it evaluates to PyObject, so just directly
-        # use Main.eval:
-        Main.eval(
-            f'Pkg.activate("{_escape_filename(processed_julia_project)}",'
-            f"shared = Bool({int(is_shared)}), "
-            f"{io_arg})"
-        )
-
-        julia_activated_env = processed_julia_project
+# Required to avoid segfaults (https://juliapy.github.io/PythonCall.jl/dev/faq/)
+if os.environ.get("PYTHON_JULIACALL_HANDLE_SIGNALS", "yes") != "yes":
+    warnings.warn(
+        "PYTHON_JULIACALL_HANDLE_SIGNALS environment variable is set to something other than 'yes' or ''. "
+        + "You will experience segfaults if running with multithreading."
+    )
 
-    if not julia_initialized:
-        julia_kwargs_at_initialization = julia_kwargs
+os.environ["PYTHON_JULIACALL_HANDLE_SIGNALS"] = "yes"
+os.environ["JULIA_NUM_THREADS"] = os.environ.get("JULIA_NUM_THREADS", "auto")
 
-    julia_initialized = True
-    if return_aux:
-        return Main, {"compiled_modules": using_compiled_modules}
-    return Main
+import juliapkg
+from juliacall import Main as jl
```

Comment:
It's recommended not to use `Main` in package code, but instead to create a new module using `juliacall.newmodule`. See here: https://juliapy.github.io/PythonCall.jl/stable/juliacall/#Getting-started.

Reply:
I did this initially but it doesn’t work for a complex reason probably specific to me.

Basically it broke the distributed tests, since the extra Julia processes — created with `addprocs` from within SymbolicRegression.jl — start in `Main` and are sent all the stuff created in `Main`. Maybe there’s a way to generalize this though?

Reply:
I also want to make it so that if the user starts `juliacall` and starts creating functions, all of those functions are available to pass in PySR.

Basically PySR is a top-level *tool* rather than a lightweight dependency, so I think it might be better to put it in `Main` by default..